### PR TITLE
Instance normalisation, configurable 1D label generator, cleanup

### PR DESCRIPTION
**Pull request by torbjorgen**

### CODE REVIEW

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 1]
```diff
@@ -163,7 +165,7 @@ def _expand_categorical_label_1d_to_2d(label: float) -> np.array:
         if not 0.0 <= label <= 1.0:
```
Comment (ChrOertlin) [inline src/rdds/lib/hdf5/hd5_data_generator.py]: this check could be abstracted away to something like:

```suggestion
        if not is_label_valid(label):
```


then 
```Python
def is_label_valid(label):
return 0.0 <= label <= 1.0
```

Not strictly necessary but it makes the code more readable, plus you can test the is_valid_function with unittests and make sure it works like you want to.

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 2]
```diff
@@ -163,7 +165,7 @@ def _expand_categorical_label_1d_to_2d(label: float) -> np.array:
...
             raise ValueError(f'Invalid value encountered, got {label}')
         multiclass_label = (1.0 - label, label)
```
Comment (ChrOertlin) [inline src/rdds/lib/hdf5/hd5_data_generator.py]:
```suggestion
        return (1.0 - label, label)
```

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 3]
```diff
@@ -38,12 +39,13 @@ def __init__(self,
...
         :param replace_nan_floats_with: Floating point value to replace NaN values
         :param load_data_into_ram: Load file content into RAM if True. Improves reading performance.
-        :raises ValueError: In case internal dataset rank is not 1
+        :param expand_1d_categorical_to_2d: Unpack a 1D categorical label [0][TN] into 2D; [1, 0][TN, TP]
         """
         self._hd5_file_path: str = hd5_file_path
         self._hd5_file = h5py.File(name=self._hd5_file_path, mode='r')
         self._data_length = int
         self._replace_nan_floats_with = replace_nan_floats_with
+        self._expand_1d_categorical_to_2d = expand_1d_categorical_to_2d
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py]: Add type hint

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 4]
```diff
@@ -181,8 +183,9 @@ def __call__(self) -> Iterator[Tuple[Union[str, float], ...]]:
                                                      output_tensor_format=[self._label])
                 if len(label) > 1:
                     raise ValueError(f'Only 1D label currently supported. Got {label}')
-                labels = self._expand_categorical_label_1d_to_2d(label[0])
-                output_vector = (output_vector, ) + (labels, )  # Add label, so (data, label) is produced
+                if self._expand_1d_categorical_to_2d:
+                    label = self._expand_categorical_label_1d_to_2d(label[0])
+                output_vector = (output_vector, ) + ((label, ), )  # Add label, so (data, label) is produced
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py]: Add type hints

File: src/rdds/lib/tf/instance_normalisation_layer.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import tensorflow as tf
+
+
+@tf.keras.saving.register_keras_serializable()  # Make sure layer is available in keras save/ load operations.
+class InstanceNormalisation(tf.keras.layers.BatchNormalization):
+
```
Comment (henrikstranneheim) [inline src/rdds/lib/tf/instance_normalisation_layer.py:6]:
```suggestion
```

File: src/tests/hdf5/test_generator.py [hunk 6]
```diff
@@ -65,9 +65,34 @@ def test_hd5_data_generator_multiple_tensors(hd5_file):
...
+    data_iter = hd5_data_generator()
+    for i, xy in enumerate(data_iter):
+        nested_tensor, labels = xy
+        # THEN expect the data to be split according to output_tensor_format, and labels are produced
+        assert isinstance(nested_tensor, tuple)
+        assert len(nested_tensor) == 2
+        assert isinstance(nested_tensor[0][0], bytes)
+        assert isinstance(nested_tensor[0][1], bytes)
+        assert isinstance(nested_tensor[1][0], float)
+        label, = labels
```
Comment (henrikstranneheim) [inline src/tests/hdf5/test_generator.py]:
```suggestion
        label, - = labels
```

File: src/tests/hdf5/test_generator.py [hunk 7]
```diff
@@ -65,9 +65,34 @@ def test_hd5_data_generator_multiple_tensors(hd5_file):
...
+    """
+    hd5_file_path, data_length = hd5_file
+    # GIVEN a HD5 data generator reading from HD5 file
+    hd5_data_generator = Hd5DataGenerator(hd5_file_path=hd5_file_path,
+                                          output_tensor_format=[['dataset0', 'dataset1'], ['dataset2']],
+                                          label='label',
+                                          expand_1d_categorical_to_2d=False)
+    # WHEN iterating over data
+    data_iter = hd5_data_generator()
+    for i, xy in enumerate(data_iter):
```
Comment (henrikstranneheim) [inline src/tests/hdf5/test_generator.py]: Is there more descriptive names here?

File: src/tests/tf/test_instance_normalisation.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+NR_FEATURES = 10
+EPOCHS = 1000
+
+
+def test_instance_normalisation():
+    """
+    Test for computing Unit Normalisation across features.
+    """
+    # GIVEN some data with n features
+    c: np.ndarray = np.random.default_rng(seed=1).random((1, 1, NR_FEATURES))  # [batch, sample, features]
```
Comment (henrikstranneheim) [inline src/tests/tf/test_instance_normalisation.py]: What does c stand for?

File: src/tests/tf/test_instance_normalisation.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+    dummy_input = tf.keras.Input((1, NR_FEATURES), dtype=tf.float32)
+    # WHEN training an instance normalisation layer
+    y: tf.Tensor = InstanceNormalisation()(dummy_input)
+    model = tf.keras.Model(dummy_input, y)
+
+    def loss_fn(y_true, y_pred) -> tf.Tensor:
+        return tf.keras.losses.mean_squared_error(y_true, y_pred)
+
+    model.compile(loss=loss_fn,
```
Comment (henrikstranneheim) [inline src/tests/tf/test_instance_normalisation.py]:
```suggestion
    model.compile(loss=tf.keras.losses.mean_squared_error(y_true, y_pred),
```

File: src/tests/tf/test_instance_normalisation.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    Test for computing Unit Normalisation across features.
+    """
+    # GIVEN some data with n features
+    c: np.ndarray = np.random.default_rng(seed=1).random((1, 1, NR_FEATURES))  # [batch, sample, features]
+    c[0, 0, 0:2] += -5  # Add some additional bias
+    c[0, 0, 2:5] += +3.5
+
+    dummy_input = tf.keras.Input((1, NR_FEATURES), dtype=tf.float32)
+    # WHEN training an instance normalisation layer
+    y: tf.Tensor = InstanceNormalisation()(dummy_input)
```
Comment (henrikstranneheim) [inline src/tests/tf/test_instance_normalisation.py]: y? Try to avoid single character variables - they are hard to understand

**Review by henrikstranneheim — COMMENTED** [review]
👍

**Review by henrikstranneheim — APPROVED** [review]

File: src/rdds/lib/hdf5/hd5_data_generator.py (hunk 2 above)
Comment (ChrOertlin) [reply src/rdds/lib/hdf5/hd5_data_generator.py]: I think we can skip the variable here and return immediatly

File: src/rdds/lib/hdf5/hd5_data_generator.py (hunk 4 above)
Comment (ChrOertlin) [inline src/rdds/lib/hdf5/hd5_data_generator.py]: add type hints

File: src/tests/hdf5/test_generator.py [hunk 11]
```diff
@@ -65,9 +65,34 @@ def test_hd5_data_generator_multiple_tensors(hd5_file):
...
 
 
-def test_hd5_with_label(hd5_file):
+def test_hd5_with_1d_label(hd5_file):
     """
-    Test for yielding labels alongside data.
+    Test for yielding labels alongside data, 1D label.
+    """
+    hd5_file_path, data_length = hd5_file
+    # GIVEN a HD5 data generator reading from HD5 file
```
Comment (ChrOertlin) [inline src/tests/hdf5/test_generator.py]:
```suggestion
    # GIVEN a HD5 data generator reading from HD5 file
    hd5_file_path, data_length = hd5_file

```

File: src/tests/hdf5/test_generator.py (hunk 7 above)
Comment (ChrOertlin) [inline src/tests/hdf5/test_generator.py]:
```suggestion
    for i, xy in enumerate(data_iter):
```
renaming i, xy into something more readable helps reviewers understand what is supposed to happen a bit more.

File: src/tests/hdf5/test_generator.py [hunk 12]
```diff
@@ -65,9 +65,34 @@ def test_hd5_data_generator_multiple_tensors(hd5_file):
...
+    for i, xy in enumerate(data_iter):
+        nested_tensor, labels = xy
+        # THEN expect the data to be split according to output_tensor_format, and labels are produced
+        assert isinstance(nested_tensor, tuple)
+        assert len(nested_tensor) == 2
+        assert isinstance(nested_tensor[0][0], bytes)
+        assert isinstance(nested_tensor[0][1], bytes)
+        assert isinstance(nested_tensor[1][0], float)
+        label, = labels
+        assert 0 <= label[0] <= 1
```
Comment (ChrOertlin) [inline src/tests/hdf5/test_generator.py]: here you could use the is_valid_label function as well and keep the code DRY

File: src/tests/hdf5/test_generator.py [hunk 13]
```diff
@@ -85,7 +110,10 @@ def test_hd5_with_label(hd5_file):
         assert isinstance(nested_tensor[0][1], bytes)
         assert isinstance(nested_tensor[1][0], float)
         label, = labels
+        # THEN expect 2D label
         assert 0 <= label[0] <= 1
+        assert 0 <= label[1] <= 1
+        assert 0 <= label[0] + label[1] <= 1.0
```
Comment (ChrOertlin) [inline src/tests/hdf5/test_generator.py]: same here

**Review by ChrOertlin — APPROVED** [review]
👍  Looks good!

File: src/tests/tf/test_instance_normalisation.py (hunk 9 above)
Comment (torbjorgen) [reply src/tests/tf/test_instance_normalisation.py]: The `loss` parameter accepts only a `callable` with signature `callable(y_true, y_pred) -> tf.Tensor`.
`y_true, y_pred` is not readily available but evaluated and passed at model runtime.
One could use `lambda` to solve this, but it's ugly compared to a stand-alone method I think @henrikstranneheim

**Review by torbjorgen — COMMENTED** [review]
* Updated according to @henrikstranneheim @ChrOertlin suggestions

**Pull request merged**